lib/private/Preview/BackgroundCleanupJob.php

```diff
@@ -28,37 +28,105 @@
+use OC\Preview\Storage\Root;
 use OCP\DB\QueryBuilder\IQueryBuilder;
+use OCP\Files\IMimeTypeLoader;
 use OCP\Files\NotFoundException;
 use OCP\Files\NotPermittedException;
+use OCP\Files\SimpleFS\ISimpleFolder;
 use OCP\IDBConnection;
 
 class BackgroundCleanupJob extends TimedJob {
 
 	/** @var IDBConnection */
 	private $connection;
 
-	/** @var Factory */
-	private $appDataFactory;
+	/** @var Root */
+	private $previewFolder;
 
 	/** @var bool */
 	private $isCLI;
 
+	/** @var IMimeTypeLoader */
+	private $mimeTypeLoader;
+
 	public function __construct(IDBConnection $connection,
-								Factory $appDataFactory,
+								Root $previewFolder,
+								IMimeTypeLoader $mimeTypeLoader,
 								bool $isCLI) {
 		// Run at most once an hour
 		$this->setInterval(3600);
 
 		$this->connection = $connection;
-		$this->appDataFactory = $appDataFactory;
+		$this->previewFolder = $previewFolder;
 		$this->isCLI = $isCLI;
+		$this->mimeTypeLoader = $mimeTypeLoader;
 	}
 
 	public function run($argument) {
-		$previews = $this->appDataFactory->get('preview');
+		foreach ($this->getDeletedFiles() as $fileId) {
+			try {
+				$preview = $this->previewFolder->getFolder($fileId);
+				$preview->delete();
+			} catch (NotFoundException $e) {
+				// continue
+			} catch (NotPermittedException $e) {
+				// continue
+			}
+		}
+	}
+
+	private function getDeletedFiles(): \Iterator {
+		yield from $this->getOldPreviewLocations();
+		yield from $this->getNewPreviewLocations();
+	}
+
+	private function getOldPreviewLocations(): \Iterator {
+		$qb = $this->connection->getQueryBuilder();
+		$qb->selectDistinct('a.name')
+			->from('filecache', 'a')
+			->leftJoin('a', 'filecache', 'b', $qb->expr()->eq(
+				$qb->expr()->castColumn('a.name', IQueryBuilder::PARAM_INT), 'b.fileid'
+			))
+			->leftJoin('a', 'filecache', 'c', $qb->expr()->eq(
+				'a.fileid', 'c.parent'
+			))
+			->where(
+				$qb->expr()->andX(
+					$qb->expr()->isNull('b.fileid'),
+					$qb->expr()->eq('a.parent', $qb->createNamedParameter($this->previewFolder->getId())),
+					$qb->expr()->eq('a.mimetype', $qb->createNamedParameter($this->mimeTypeLoader->getId('httpd/unix-directory'))),
+					$qb->expr()->neq('c.mimetype', $qb->createNamedParameter($this->mimeTypeLoader->getId('httpd/unix-directory')))
+				)
+			);
+
+		if (!$this->isCLI) {
+			$qb->setMaxResults(10);
+		}
+
+		$cursor = $qb->execute();
+
+		while ($row = $cursor->fetch()) {
+			yield $row['name'];
+		}
 
-		$previewFodlerId = $previews->getId();
+		$cursor->closeCursor();
+	}
+
+	private function getNewPreviewLocations(): \Iterator {
+		$qb = $this->connection->getQueryBuilder();
+		$qb->select('path', 'mimetype')
+			->from('filecache')
+			->where($qb->expr()->eq('fileid', $qb->createNamedParameter($this->previewFolder->getId())));
+		$cursor = $qb->execute();
+		$data = $cursor->fetch();
+		$cursor->closeCursor();
+
+		if ($data === null) {
+			return [];
+		}
+
+		$like = $data['path'] . '/_/_/_/_/_/_/_/%';
```

Comment:
You might want to escape `$data['path']`
```suggestion
		$like = $this->connection->escapeLikeParameter($data['path']) . '/_/_/_/_/_/_/_/%';
```

Reply:
fair enough